### PR TITLE
Update Build and Testing to use Bazel

### PR DESCRIPTION
**Pull request by srvasude**
Few different things in this PR:

* Made this project a bazel workspace, and added BUILD rules for most things in cxx/. In particular, I didn't add BUILD rules for two of the tests in tests/ because I see them being deleted soon (subsumed by other tests we have written / we will write).

* To import boost, I added a github repo to MODULE.lock that provides bazel bindings to boost. I think this is fine, and this has worked locally.

* I've had to update a bunch of cc files because bazel was catching things ranging from lint to actual errors.

* I've updated the makefile and kept the make system here. I'll delete it in a later change, once we have scripts for running various things.

### CODE REVIEW

File: cxx/MODULE.bazel.lock
Comment (emilyfertig) [inline cxx/MODULE.bazel.lock:1]: Maybe we should gitignore this? It looks like it's auto-generated and can cause merge conflicts (though they fixed it at head: https://github.com/bazelbuild/bazel/issues/20369)

**Review by emilyfertig — APPROVED** [review]
Thank you!

File: cxx/MODULE.bazel.lock
Comment (srvasude) [reply cxx/MODULE.bazel.lock:1]: Ah meant to delete from this change (added to .gitignore).

**Comment by srvasude** [comment]
Just mentioning things I tested (for posterity, since we may catch some issue for changing build systems):

make integration_tests (these ran successfully and printed out the right things).
blaze build :all in different directories
blaze test :all in the different directories

**Pull request merged**